code/modules/integrated_electronics/subtypes/mining.dm

```diff
@@ -0,0 +1,103 @@
+/obj/item/integrated_circuit/mining/ore_analyzer
+	name = "ore analyzer"
+	desc = "Analyzes a rock for its ore type."
+	extended_desc = "Takes a reference for an object and checks if it is a rock first. If that is the case, it outputs the mineral \
+	inside the rock."
+	category_text = "Mining"
+	ext_cooldown = 1
+	complexity = 20
+	cooldown_per_use = 1 SECONDS
+	inputs = list(
+		"rock" = IC_PINTYPE_REF
+		)
+	outputs = list(
+		"ore type" = IC_PINTYPE_STRING,
+		"amount" = IC_PINTYPE_NUMBER
+		)
+	activators = list(
+		"analyze" = IC_PINTYPE_PULSE_IN,
+		"on analyzed" = IC_PINTYPE_PULSE_OUT,
+		"on found" = IC_PINTYPE_PULSE_OUT
+		)
+	spawn_flags = IC_SPAWN_RESEARCH
+	power_draw_per_use = 20
+	var/turf/closed/mineral/rock
+	var/mineral
+
+
+/obj/item/integrated_circuit/mining/ore_analyzer/do_work()
+	rock = get_pin_data(IC_INPUT, 1)
+	if(!istype(rock,/turf/closed/mineral))
+		mineral = null
+		set_pin_data(IC_OUTPUT, 1, null)
+		set_pin_data(IC_OUTPUT, 2, null)
+
+		push_data()
+		activate_pin(2)
+		return
+
+	if(rock.mineralType)
+		mineral = "[rock.mineralType]"
+		mineral = copytext(mineral, findlasttextEx(mineral, "/")+1)
+
+	else
+		mineral = "rock"
+
+	set_pin_data(IC_OUTPUT, 1, mineral)
+	set_pin_data(IC_OUTPUT, 2, rock.mineralAmt)
+	push_data()
+
+	activate_pin(2)
+	activate_pin(3)
+
+
+/obj/item/integrated_circuit/mining/mining_drill
+	name = "mining drill"
+	desc = "A mining drill that can drill through rocks."
+	extended_desc = "A mining drill that activates on sensing a mineable rock. It takes some time to get the job done and \
+	has to not be moved during that time."
+	category_text = "Mining"
+	ext_cooldown = 1
+	complexity = 20
+	cooldown_per_use = 3 SECONDS
+	inputs = list(
+		"rock" = IC_PINTYPE_REF
+		)
+	outputs = list()
+	activators = list(
+		"mine" = IC_PINTYPE_PULSE_IN,
+		"on success" = IC_PINTYPE_PULSE_OUT,
+		"on failure" = IC_PINTYPE_PULSE_OUT
+		)
+	spawn_flags = IC_SPAWN_RESEARCH
+	power_draw_per_use = 100
+
+	var/busy = FALSE
+	var/usedx
+	var/usedy
+	var/turf/closed/mineral/rock
+
+/obj/item/integrated_circuit/mining/mining_drill/do_work()
+	rock = get_pin_data(IC_INPUT, 1)
+
+	if(!(ismineralturf(rock) && (get_dist(assembly,rock) < 2)) || busy)
```

Comment:
`(get_dist(assembly,rock) < 2)`

Could this not be r`ock.Adjacent(assembly)`

Reply:
Has some issues when you are doing it from pocket or hand, would work well on a mmi-bot